presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -216,7 +226,7 @@ private static boolean canCastTypeBase(String fromTypeBase, String toTypeBase)
         }
         switch (fromTypeBase) {
             case StandardTypes.BIGINT:
-                return StandardTypes.DOUBLE.equals(toTypeBase);
+                return StandardTypes.DOUBLE.equals(toTypeBase) || StandardTypes.DECIMAL.equals(toTypeBase);
             case StandardTypes.DATE:
```

Comment:
Why sometimes we return false and sometimes we throw IllegalStateException. This seem quite unnatural to me.